crates/matrix-sdk-sled/src/state_store.rs

```diff
@@ -149,8 +173,131 @@ const TIMELINE_METADATA: &str = "timeline-metadata";
 #[cfg(feature = "experimental-timeline")]
 const TIMELINE: &str = "timeline";
 
+const ALL_DB_STORES: &[&str] = &[
+    ACCOUNT_DATA,
+    SYNC_TOKEN,
+    DISPLAY_NAME,
+    INVITED_USER_ID,
+    JOINED_USER_ID,
+    MEDIA,
+    MEMBER,
+    PRESENCE,
+    PROFILE,
+    ROOM_ACCOUNT_DATA,
+    #[cfg(feature = "experimental-timeline")]
+    ROOM_EVENT_ID_POSITION,
+    ROOM_EVENT_RECEIPT,
+    ROOM_INFO,
+    ROOM_STATE,
+    ROOM_USER_RECEIPT,
+    ROOM,
+    SESSION,
+    STRIPPED_INVITED_USER_ID,
+    STRIPPED_JOINED_USER_ID,
+    STRIPPED_ROOM_INFO,
+    STRIPPED_ROOM_MEMBER,
+    STRIPPED_ROOM_STATE,
+    CUSTOM,
+    #[cfg(feature = "experimental-timeline")]
+    ROOM_EVENT_ID_POSITION,
+    #[cfg(feature = "experimental-timeline")]
+    TIMELINE_METADATA,
+    #[cfg(feature = "experimental-timeline")]
+    TIMELINE,
+];
+const ALL_GLOBAL_KEYS: &[&str] = &[VERSION_KEY];
+
 type Result<A, E = SledStoreError> = std::result::Result<A, E>;
 
+#[derive(Builder, Debug, PartialEq, Eq)]
+#[builder(name = "SledStoreBuilder", build_fn(skip))]
+pub struct SledStoreBuilderConfig {
+    path: PathBuf,
+    passphrase: String,
+    #[builder(default = "MigrationConflictStrategy::BackupAndDrop")]
+    migration_conflict_strategy: MigrationConflictStrategy,
+}
+
+impl SledStoreBuilder {
+    pub fn build(&mut self) -> Result<SledStore> {
+        let is_temp = self.path.is_none();
+
+        let mut cfg = Config::new().temporary(is_temp);
+
+        let path = if let Some(path) = &self.path {
+            let path = path.join("matrix-sdk-state");
+
+            cfg = cfg.path(&path);
+            Some(path)
+        } else {
+            None
+        };
```

Comment:
Can be `self.path.as_ref().map(...)`.

Reply:
it doesn't as it also modified `cfg` and putting that into a `map` means it is moved, but then needed later. sure there is a way to do that, but it becomes icky...

Reply:
No, closures are by-ref unless you use the `move` keyword. [Demo](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fd65b79903dfe3a1a0ac049b5282f590)

Reply:
I don't insist.

Reply:
hmmm... the demo doesn't seem applicable here... maybe because `string` is `copy`?.

When I do:
```rust
        let mut cfg = Config::new().temporary(is_temp);

        let path = self.path.as_ref().map(|origin_path| {
            let path = origin_path.join("matrix-sdk-state");

            cfg = cfg.path(&path);
            path
        });

        let db = cfg.open().map_err(StoreError::backend)?;

```

The compiler tells me:

```
warning: value assigned to `cfg` is never read
   --> crates/matrix-sdk-sled/src/state_store.rs:232:13
    |
232 |             cfg = cfg.path(&path);
    |             ^^^
    |
    = note: `#[warn(unused_assignments)]` on by default
    = help: maybe it is overwritten before being read?

error[E0382]: borrow of moved value: `cfg`
   --> crates/matrix-sdk-sled/src/state_store.rs:236:18
    |
227 |         let mut cfg = Config::new().temporary(is_temp);
    |             ------- move occurs because `cfg` has type `Config`, which does not implement the `Copy` trait
228 | 
229 |         let path =self.path.as_ref().map(|origin_path| {
    |                                          ------------- value moved into closure here
...
232 |             cfg = cfg.path(&path);
    |                   --- variable moved due to use in closure
...
236 |         let db = cfg.open().map_err(StoreError::backend)?;
    |                  ^^^^^^^^^^ value borrowed here after move

For more information about this error, try `rustc --explain E0382`.
warning: `matrix-sdk-sled` (lib) generated 1 warning
error: could not compile `matrix-sdk-sled` due to previous error; 1 warning emitted
```

Reply:
(it's also what I tried before, when I made the previous comment)

Reply:
No, `String` is not `Copy`. I'll try to understand what's going on here, but I don't want to block this PR on it.